PLT: Cleaner plotting backend API, and unify Series and DataFrame accessors

datapythonista opened the pull request:
- [X] xref #26747
- [ ] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

The main changes are:

The API that backends need to implement is simply a function `plot(data, kind, **kwargs)`. After testing different approaches (abstract class previously mentioned and others), I think this is what makes things simpler in both ends pandas and the backends. For `hvplot` `hvPlot.__call__` is already what needs to be implemented, and this change will mainly require to delete a bit of code. For `pdvega` a bit more work will be needed, but this will help remove a significant amount of repetitive code.

Second thing, more in the pandas side, I unify the accessors of `Series` and `DataFrame`. There will be minor changes, like unified documentation (which I think it's great, but docstrings will have to be generic), and also `Series.plot.scatter` will exist (which now doesn't) but will generate an exception if called. Not a big deal I'd say, because now it's already possible to do `Series.plot(kind='scatter')`, so it'll just make both cases consistent. Same things apply to `hexbin`.

CC: @pandas-dev/pandas-core @philippjfr @jakevdp

datapythonista committed:
Some experiments so far

datapythonista committed:
Refactoring of pandas plotting to make the API clearer

TomAugspurger commented:
I'm still a bit unclear on how the dispatching to backends will work. IIUC, on master we structure things as `DataFrame.plot.__call__(kind, ...)`, which selects the right method based on the `kind`. So `DataFrame.plot(kind='line')` calls `DataFrame.plot.line` which does its own thing.

Now we're going to change things a bit so that `DataFrame.plot.line` calls `DataFrame.plot.__call__` with `kind=line`.

Does that have a direct impact on how backends will provide their own implementations?

datapythonista commented:
> Now we're going to change things a bit so that `DataFrame.plot.line` calls `DataFrame.plot.__call__` with `kind=line`.

Not exactly. I didn't modify that part (besides merging the accessors for Series and DataFrame). In master `DataFrame.plot.line` calls the accessor `line` method, and the `line` method calls `__call__`. And this didn't change.

The `__call__` method was calling `plot_frame` and `plot_series`, and those were calling `plot`. `plot was instantiating/calling `LinePlot()` (`LinePlot` was the public part of the backend). This is the part that changed.

Now `__call__` will call `plot` in the backend (`plot` is the public part of the backend). And `plot` will instantiate and call `LinePlot` and the other classes. I think this is much better, since all backends will run some common code first no matter the plot kind.

datapythonista commented on pandas/plotting/_core.py:
```diff
@@ -630,389 +31,133 @@ def _get_plot_backend():
     ``s.plot(kind='line')`` is equivalent to ``s.plot.line()``
     """
+    def __init__(self, data):
+        assert isinstance(data, (pandas.Series, pandas.DataFrame))
```
There was a comment saying that `data` had to be `Series` or `DataFrame`, I thought that an assert would say the same and also check. No other reason, and happy to get rid of it.

datapythonista commented on pandas/plotting/_matplotlib/__init__.py:
```diff
@@ -13,13 +14,37 @@
-           'bootstrap_plot', 'lag_plot', 'parallel_coordinates', 'radviz',
-           'scatter_matrix', 'register', 'deregister']
+def plot(data, kind, **kwargs):
+    if isinstance(data, pandas.Series):
```
Makes sense; not sure if copied this, but will change it.

Commit:
Merge remote-tracking branch 'upstream/master' into plot_api

Commit:
Addressing review comments, and fixing many tests (still some tests failing)

Commit:
Merge remote-tracking branch 'upstream/master' into plot_api

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -1667,8 +956,123 @@ def hexbin(self, x, y, C=None, reduce_C_function=None, gridsize=None,
+                                    figsize=figsize, bins=bins, **kwds)
+
+
+hist_series.__doc__ = PlotAccessor.hist.__doc__
```
I would leave the original hist_series and hist_frame intact, as the accessor's docstring version is not correct for them (different keywords, different order of the keywords, ...)

jorisvandenbossche replied:
And the same for boxplot below

datapythonista replied:
Didn't realize, good point, thanks!

jorisvandenbossche reviewed (commented):
As said on the issue, +1 on such a simpler functional API!

Commit:
Merge remote-tracking branch 'upstream/master' into plot_api

Commit:
Restoring docstrings of hist_series, hist_frame, boxplot and boxplot_frame

Commit:
Fixing plot accessor docstring (was in the wrong place, and couple of things were wrong or could be improved)

Commit:
Fixing hexbin plot tests

Commit:
Fixing bug when calling plot twice on the same data, since the data (Series or DataFrame) was being modified in the first call

Commit:
Raising missing exception for pie in DataFrame, and fixing accessor signature mismatches

Commit:
Fixing bug that shown the legend for Series plot

Commit:
Fix linting

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -370,1305 +642,760 @@
-        Plot a whole dataframe to a bar plot. Each column is assigned a
-        distinct color, and each row is nested in a group along the
-        horizontal axis.
+def hist_frame(data, column=None, by=None, grid=True, xlabelsize=None,
```
It makes sense to move those non-accessor functions down in the file, but for the (already large) diff it might make sense to keep them at the top where they were (those ones are just a copy paste I think), and only do the move in a follow-up PR

datapythonista replied:
Yes, that makes sense, but surely better in a follow up.

jorisvandenbossche replied:
@datapythonista I actually meant moving them back here, so we can do the actual move in a follow-up 
(the diff is right now basically not informative)

Commit:
Merge remote-tracking branch 'upstream/master' into plot_api

datapythonista commented:
Still working on the failing tests here. Of the two that are failing locally, the problem can be reproduced with this simple example:
```python
import pandas

df = pandas.DataFrame({'a': [.19, .78, .27],
                       'b': [.62, .77, .80]})
df.plot()
df.plot(y='a')
```

In master this generates two plots, but in this branch is generating a single plot for both calls. Couldn't figure out why yet.

jakevdp commented:
> In master this generates two plots, but in this branch is generating a single plot for both calls.

I think it's because on master, plot_frame lacks the check for an existing figure: ``if ax is None and len(plt.get_fignums()) > 0``

After this change, `df.plot` does have this code.

jorisvandenbossche commented:
> I'm not certain whether this difference in behavior between dataframe plotting and series plotting is intentional.

@jakevdp It's in any case a long standing behaviour (with quite some disagreement which of the two is better): https://github.com/pandas-dev/pandas/issues/8776

jorisvandenbossche commented:
@datapythonista do you have time the coming days to try to fix the tests?

toobaz commented on pandas/plotting/_core.py:
```diff
@@ -370,1305 +642,760 @@
+        if not isinstance(self._parent,
+                          pandas.core.dtypes.generic.ABCDataFrame):
+            raise ValueError(
+                "plot kind scatter can only be used for data frames")
```
Doesn't this duplicate the check at https://github.com/pandas-dev/pandas/pull/27009/files#diff-2b118bda866f3d626ceb6b529c62cd1aR222 ?

toobaz commented on pandas/plotting/_core.py:
```diff
@@ -370,1305 +642,760 @@
+        if not isinstance(self._parent,
+                          pandas.core.dtypes.generic.ABCDataFrame):
+            raise ValueError(
+                "plot kind hexbin can only be used for data frames")
```
Same comment as above (duplicated check?)

toobaz reviewed (commented):
Two small comments, but great job!

Commit:
Merge remote-tracking branch 'upstream/master' into plot_api

Commit:
Fixing bug that made reusing the previous plot for dataframes

Commit:
Removing duplicated data type checks

Commit:
Restoring original position of methods, so the diff is smaller

datapythonista commented:
Thanks for the reviews, and thanks a lot for the comment @jakevdp, didn't realize of that behavior, but that made me understand the problem, I was very confused.

The problem was that I moved the code making the plot be in the same figure for Series to the backend, and the backend may receive a Series for DataFrame plots since we convert the data in some cases. I added an extra parameter `new_plot`  for the backend. It'll surely make sense to continue the discussion on #8776, but not making any functional change in this refactoring.

@jorisvandenbossche I kept the original order of the methods in this PR as you suggested, but looks like git is not smart enough in this case and the diff is still huge.

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -597,103 +397,227 @@ def boxplot_frame_groupby(grouped, subplots=True, column=None, fontsize=None,
+        data = self._parent.copy()
+
+        if isinstance(data, pandas.core.dtypes.generic.ABCSeries):
+            kwargs['new_plot'] = True
```
This sounds a bit strange (for Series I except to get *no* new plot, so False), so should the meaning of True/False be the other way around?

datapythonista replied:
yes, you're right, I started to implement for `new_plot`, and changed it later and forgot to rename to `reuse_plot`

jorisvandenbossche commented:
> @jorisvandenbossche I kept the original order of the methods in this PR as you suggested, but looks like git is not smart enough in this case and the diff is still huge.

Sorry for letting you do this then, I hoped git was smarter :) (but I think it did *something*)

datapythonista commented:
No worries, I think the change made sense, but I guess there are too many lines deleted at the beginning of the file for git to match correctly.

There are couple of tests failing with matplotlib 2.2, but other than that things look good now.

Commit:
Fixing name of reuse_plot parameter

Commit:
Fixing bug with matplotlib 2

datapythonista commented:
All green here finally. :)

I think all comments from reviews have been addressed too, let me know if there is anything else.

TomAugspurger reviewed (approved):
LGTM. @jorisvandenbossche did you have anything else?

jreback commented on pandas/plotting/_core.py:
```diff
@@ -18,160 +17,148 @@
-    `**kwds` : keywords
-        Options to pass to matplotlib plotting method
+
+def hist_frame(data, column=None, by=None, grid=True, xlabelsize=None,
```
at some point in time should type these

jreback commented on pandas/plotting/_core.py:
```diff
@@ -597,103 +397,227 @@ def boxplot_frame_groupby(grouped, subplots=True, column=None, fontsize=None,
+    Make plots of Series or DataFrame using the backend specified by the
+    option ``plotting.backend``. By default, matplotlib is used.
 
+    *New in version 0.17.0:* Each plot kind has a corresponding method on
```
should fix these types of coments (and user versionadded and such); though this particular one can drop

jreback commented on pandas/plotting/_core.py:
```diff
@@ -597,103 +397,227 @@ def boxplot_frame_groupby(grouped, subplots=True, column=None, fontsize=None,
+        We used to have different accessors for Series and DataFrame. Their
+        signatures were different:
+
+        - SeriesPlotMethods.__call__(kind, ..., **kwargs)
```
this is not a very helpful comment above here; focus on what a future reader will see (what you say below)

jreback commented on pandas/plotting/_core.py:
```diff
@@ -597,103 +397,227 @@ def boxplot_frame_groupby(grouped, subplots=True, column=None, fontsize=None,
+        This function makes this unified `__call__` method compatible with both
+        """
+        if args and isinstance(data, ABCSeries):
+            # TODO raise warning here, positional arguments shouldn't be
```
can you add this raise? do we test this? alt if you'd make an issue an reference it here, we don't want to have TODOs w/o an issue.

jreback commented on pandas/plotting/_core.py:
```diff
@@ -597,103 +397,227 @@ def boxplot_frame_groupby(grouped, subplots=True, column=None, fontsize=None,
+
+        plot_backend = _get_plot_backend()
+        data = self._parent.copy()
+
```
I would add some helpful comments here

jreback reviewed (changes requested):
looks good, just a couple of small comments. Can you put a comments section somewhere at the top-level explaining what the interface for this looks like (meaning if you are an author of a library and  you want to implement the interface), what do you have to do (sketch out the code); could also be in a developing.rst doc, but I think needs *some* things here.

Commit:
Adding documentation and improving comments, based on Jeff review

Commit:
Adding FutureWarning if Series.plot is called with positional arguments

Commit:
Not passing default matplotlib parameters to backends (all known kwargs were being passed, now only the ones provided by the user)

Commit:
Fixing test of plotting accessor parameters

Commit:
Temporary not warning for Series.plot positional arguments (looks like this caused the parallel coordinates test failure)

pep8speaks commented:
Hello @datapythonista! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:







* In the file [`pandas/plotting/_core.py`](https://github.com/pandas-dev/pandas/blob/4d70d5d98bf71649c438ed08ce339169d3f3aeb7/pandas/plotting/_core.py):

> [Line 549:80](https://github.com/pandas-dev/pandas/blob/4d70d5d98bf71649c438ed08ce339169d3f3aeb7/pandas/plotting/_core.py#L549): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (149 > 79 characters)

Commit:
Revert "Temporary not warning for Series.plot positional arguments (looks like this caused the parallel coordinates test failure)"

This reverts commit 4d70d5d98bf71649c438ed08ce339169d3f3aeb7.

Commit:
Merge remote-tracking branch 'upstream/master' into plot_api

jorisvandenbossche commented on pandas/plotting/__init__.py:
```diff
@@ -1,18 +1,73 @@
+future (e.g. `title` which should be useful for any backend).
+
+Currently, all the Matplotlib functions in pandas are accessed through
+the selected backend. For example, `pandas.plotting.boxplot` (equivalent
```
I don't think we should mention those right now, as there is still some discussion on whether to see them as part of the plotting backend interface or not.

datapythonista replied:
two lines below I say that `This is expected to change`. Personally I think it adds more value to have this comment now, and remove it when/if that changes, than just not having this documented, or document the future behavior.

Commit:
Adding debug info in the CI for failing test

jorisvandenbossche commented:
@datapythonista how are the changes in this PR affecting the misc plots? (since the failing tests seem to be related to those)

datapythonista commented:
No idea, I think the test started to fail after some changes to the documentation and comments, or a merge, doesn't make any sense. The changes here shouldn't affect andrews_curves at all. Can't reproduce locally, so I need to debug in the CI.

Commit:
Revert "Adding debug info in the CI for failing test"

This reverts commit 34d189f6b7c9ff03d72e9c382115db9f2e6074aa.

Commit:
Temporary removing the warning, to see if it's causing the andrews_curves test failure

datapythonista commented:
Failing example:
```python
from pandas import DataFrame
from pandas.plotting import parallel_coordinates

df = DataFrame({"feat": [i for i in range(30)],
                "class": [2 for _ in range(10)] +
                         [3 for _ in range(10)] +
                         [1 for _ in range(10)]})
ax = parallel_coordinates(df, 'class', sort_labels=True)
polylines, labels = ax.get_legend_handles_labels()
color_label_tuples = \
    zip([polyline.get_color() for polyline in polylines], labels)
ordered_color_label_tuples = sorted(color_label_tuples,
                                    key=lambda x: x[1])
ordered_color_label_tuples
```

Locally returns:
```
[([0.417022004702574, 0.7203244934421581, 0.00011437481734488664], '1'),
 ([0.43599490214200376, 0.025926231827891333, 0.5496624778787091], '2'),
 ([0.5488135039273248, 0.7151893663724195, 0.6027633760716439], '3')]
```
Which is correct, but in the Mac build in the CI seems to return:
```
[([0.417022004702574, 0.7203244934421581, 0.00011437481734488664], '1'),
 ([0.43599490214200376, 0.025926231827891333, 0.5496624778787091], '2'),
 ([0.5488135039273248, 0.7151893663724195, 0.6027633760716439], '3'),
 ('#1f77b4', 'None')]
```
No idea where this mysterious last element comes from. If anyone with a Mac can reproduce it locally that can makes things easy.

The CI was green, and this started to fail after updating comments and the docs, and adding a warning for `Series.plot` with positional arguments. I'm temporary removing the warning to see if it is causing the error.

jorisvandenbossche commented:
>  I'm temporary removing the warning to see if it is causing the error.

It didn't help .. (the mac build still has the additional parallel_coordinates error).

I am fine with skipping that test on Mac to get this merged, and try to fix in a follow-up.

datapythonista commented:
Very weird, the other commit when this started to fail was just on comments and docs. Will continue tomorrow.

jreback committed:
Merge branch 'master' into PR_TOOL_MERGE_PR_27009

datapythonista committed:
Revert "Temporary removing the warning, to see if it's causing the andrews_curves test failure"

This reverts commit 29d7547ce49c0f4afb0296237e43ad2455f9b6c7.

datapythonista commented:
Looks like this is the commit that is making the parallel_coordinates test fail: https://github.com/pandas-dev/pandas/pull/27204/commits/16d1f88cfcb7b30f86637422f550db4c50d58185

I'll remove it for now if that's all right.

datapythonista committed:
Removing test that causes parallel_coordinates test to fail

jorisvandenbossche commented:
> Looks like this is the commit that is making the parallel_coordinates test fail: 16d1f88

It might be that putting it *in* the class fixes it. The setup method of that class imports matplotlib (which might influence if that is imported before pyplot).

datapythonista commented:
The warning is tested in another test anyway, it's redundant, so I think it's not a problem to remove that test.

@jreback you requested changes, can you have a look?

jreback commented:
thanks @datapythonista very nice!

prob want to have a look at built docs / doc-strings to make sure everything renderning correctly.

TomAugspurger commented:
Nice work @datapythonista! Anything else you want to get to for the release candidate? I'm hoping to tag it in ~7 hours.

jorisvandenbossche commented:
I just commented on the issue that I think we should decide about the scope of the plotting backend (include misc plotting functions or not) for 0.25.0, but that doesn't need to be done for the RC.

datapythonista commented:
will check the docs once they're built.

Thanks Tom, we started to work on hvplot, and I fixed here couple of minor things that were creating problems for the backends. I think everything should be all right on this now, but if there is anything should be very small and can wait after the RC.

The main thing pending on this affecting the release is what Joris just mentioned on what we delegate to the backends. We can discuss that in #26747.
